### PR TITLE
Pencilblue 0.9.0

### PR DESCRIPTION
**Pull request by lgenet**
This is it.  Pencilblue 1.0.  A complete re-write of many of the core systems. breaking changes and deprecation removal.  The system is now backed by KAO and will be completely Async/Await and ES6 Complaint.  

Watch this space to see as the progress continues.

### CODE REVIEW

File: controllers/base_controller.js [hunk 1]
```diff
@@ -652,10 +593,12 @@ module.exports = function BaseControllerModule(pb) {
      * @param {String} location
      * @param {Function} cb
      */
-    BaseController.prototype.redirect = function(location, cb){
-        cb(pb.RequestHandler.generateRedirect(location));
+    BaseController.prototype.redirect = function (location, cb) {
+        cb(pb.Errors.redirect(location));
+    };
+    BaseController.prototype.doRedirect = function (destination, code = 302) {
+        throw pb.Errors.redirect(destination, code);
```
Comment (josesubi) [inline controllers/base_controller.js:600]: If I understand correctly you are returning and throwing an error here so nobody should use these methods any more  , why you don't add a comment or meta data to show that the methods should not be used any more and which should be the proper way to perform the redirection, and also I see a lot of references in Pb to this redirect

Comment (lgenet) [reply controllers/base_controller.js:600]: So we have a minor issue in the way of KOA and its handling of redirection as integrated with Pencilblue.  The only way to get a redirect to happen is to force the middleware stack to end.  The only way to force the middleware stack to end is to throw an error (or not call next, but that is a bit more complicated given the safe guards in the system around isolating plugin code from system code.) 

So this was the approach that was taken.  Some more work can be done around handling and cleaning this up.  like setting a - flag on the body that will stop execution of the middleware stack.

File: include/index.js [hunk 2]
```diff
@@ -211,9 +214,8 @@ module.exports = function PB(config) {
     pb.AdminSubnavService = require(path.join(config.docRoot, '/include/service/admin/admin_subnav_service.js'))(pb);
     pb.AnalyticsManager   = require(path.join(config.docRoot, '/include/system/analytics_manager.js'))(pb);
     pb.UrlService         = require(path.join(config.docRoot, '/include/service/entities/url_service.js'))(pb);
-    pb.CallHomeService    = require(path.join(config.docRoot, '/include/system/call_home_service.js'))(pb);
     pb.JobService         = require(path.join(config.docRoot, '/include/service/entities/job_service.js'))(pb);
-    pb.TokenService       = require(path.join(config.docRoot, '/include/service/entities/token_service.js'))(pb);
+    pb.TokenServiceV2     = require(path.join(config.docRoot, '/include/koa/authentication/token_service.js'))(pb);
```
Comment (DanielCoello) [inline include/index.js:219]: If we remove the TokenService, we need to update the AuthTokenControllerModule(plugins/pencilblue/controllers/actions/admin/sites/auth_token.js). That is the one that manage the session (connect to another site when you're logged in global). I tested it out right now, and it throws the following error:

`2018-11-16T22:17:36.862Z - error: [M]  TypeError: pb.TokenService is not a constructor
    at AuthTokenControllerModule.AuthTokenController.render (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/plugins/pencilblue/controllers/actions/admin/sites/auth_token.js:40:28)
    at Promise (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/include/http/middleware/controller.js:27:48)
    at Promise._execute (/Users/dacm26/Documents/Projects/Number8/career-builder/CMSPencilblue/node_modules/bluebird/js/release/debuggability.js:303:9)
    at Promise._resolveFromExecutor (/Users/dacm26/Documents/Projects/Number8/career-builder/CMSPencilblue/node_modules/bluebird/js/release/promise.js:483:18)
    at new Promise (/Users/dacm26/Documents/Projects/Number8/career-builder/CMSPencilblue/node_modules/bluebird/js/release/promise.js:79:10)
    at render (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/include/http/middleware/controller.js:26:29)
    at dispatch (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/node_modules/koa-router/node_modules/koa-compose/index.js:44:32)
    at next (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/node_modules/koa-router/node_modules/koa-compose/index.js:45:18)
    at setSecureResponseHeaders (/Users/dacm26/Documents/Projects/Number8/career-builder/CMSPencilblue/plugins/cms_base/middleware/index.js:78:20)
    at dispatch (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/node_modules/koa-router/node_modules/koa-compose/index.js:44:32)
    at next (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/node_modules/koa-router/node_modules/koa-compose/index.js:45:18)
    at initializeController (/Users/dacm26/Documents/Projects/Number8/career-builder/pencilblue/include/http/middleware/controller.js:19:15)
    at <anonymous>:null:null`

File: include/http/middleware/controller.js [hunk 3]
```diff
@@ -1,40 +1,53 @@
...
-    initializeController: (req, res) => {
-        const props = pb.RequestHandler.buildControllerContext(req, res, {});
-        return new Promise((resolve, reject) => {
-            req.controllerInstance.init(props, (err) => {
+    initializeController: async (ctx, next) => {
+        const props = pb.RequestHandler.buildControllerContext(ctx);
+        await new Promise((resolve, reject) => {
+            ctx.req.controllerInstance.init(props, (err) => {
                 if (util.isError(err)) {
                     return reject(err)
```
Comment (JoseDomingoMoraNunezCB) [inline include/http/middleware/controller.js:14]: Minor: add semicolons to these lines to follow the pattern

File: include/service/entities/site_service.js [hunk 4]
```diff
@@ -121,9 +125,9 @@ module.exports = function SiteServiceModule(pb) {
      * @method getAllSites
      * @param {Function} cb - the callback function
      */
-    SiteService.prototype.getAllSites = function (cb) {
-        var dao = new pb.DAO();
-        dao.q(SITE_COLL, { select: pb.DAO.PROJECT_ALL, where: {} }, cb);
+    SiteService.prototype.getAllSites = async function () {
+        let dao = new pb.DAO();
```
Comment (JoseDomingoMoraNunezCB) [inline include/service/entities/site_service.js:129]: Why you don't use this.dao here?

File: controllers/base_controller.js [hunk 5]
```diff
@@ -151,19 +154,20 @@ module.exports = function BaseControllerModule(pb) {
...
          */
-        this.localizationService = props.localization_service;
+        this.localizationService = props.localization_service || {language: 'en-US', g: (key) => key};
 
         /**
          * @property ls
          * @type {Localization}
          */
-        this.ls = props.localization_service;
+        this.ls = props.localization_service || {language: 'en-US', g: (key) => key};
```
Comment (bradleylowenstein) [inline controllers/base_controller.js:170]: Perhaps just have this.ls = this.localizationService

Comment (bradleylowenstein) [reply controllers/base_controller.js:170]: Scratch this comment as this.localizationService will be removed

File: controllers/base_controller.js [hunk 6]
```diff
@@ -475,58 +458,16 @@ module.exports = function BaseControllerModule(pb) {
...
-                cb(err, null);
-            }
-        });
-        this.req.on('end', function () {
-
-            //create one big buffer.
-            var body = Buffer.concat (buffers, totalLength);
-            cb(null, body);
-        });
+        cb(null, this.body);
```
Comment (bradleylowenstein) [inline controllers/base_controller.js:470]: There are a few methods calling this cb(null, this.body).
Can we have one method to return this response so if this changes again, only one place to change the response?

File: controllers/base_controller.js [hunk 7]
```diff
@@ -652,10 +593,12 @@ module.exports = function BaseControllerModule(pb) {
      * @param {String} location
      * @param {Function} cb
      */
-    BaseController.prototype.redirect = function(location, cb){
-        cb(pb.RequestHandler.generateRedirect(location));
+    BaseController.prototype.redirect = function (location, cb) {
+        cb(pb.Errors.redirect(location));
```
Comment (bradleylowenstein) [inline controllers/base_controller.js:597]: Please add some comments on what this is doing for future changes.

File: include/config.js [hunk 8]
```diff
@@ -318,7 +318,7 @@ Configuration.getBaseConfig = function(multisite) {
             //be set to "auto".  This will instruct the master process to inspect
             //the number of cores on the server and spawn a child process for each
             //core.
-            workers: 1,
+            workers: 5,
```
Comment (lgenet) [inline include/config.js:321]: Put this back to 1.

Comment (lgenet) [reply include/config.js:321]: also remove main, new main, and the other junk files

File: include/http/router.js [hunk 9]
```diff
@@ -1,297 +1,197 @@
...
+        constructor() {
+            this.app = new Koa();
+            this.app.keys = ['9011fa34-41a6-4a4d-8ad7-d591c5d3ca01']; // Random GUID
+            this.app._requestsServed = this.app._requestsServed || 0;
+
+            this.router = new Router();
+
+            this.app.use(bodyParser({
+                multipart: true,
+                // formidable: { uploadDir: path.join(__dirname, 'tmp') }
```
Comment (lgenet) [inline include/http/router.js:23]: remove commented code

File: include/http/router.js [hunk 10]
```diff
@@ -1,297 +1,197 @@
...
-         * @returns {number} The position of the middleware or -1 when not found
+        startSSLServer(port) {
+            const config = {
+                https: {
+                    port,
+                    options: {
+                        key: fs.readFileSync(pb.config.server.ssl.key, 'utf8'),
+                        cert: fs.readFileSync(pb.config.server.ssl.cert, 'utf8')
+                    }
+                }
```
Comment (lgenet) [inline include/http/router.js:150]: Hey logan, this is random but you broken token service.  Fix that.

File: include/http/middleware/controller.js (hunk 3 above)
Comment (lgenet) [reply include/http/middleware/controller.js:14]: 👍

**Pull request merged**